src/launcher/plugin.ts

```diff
@@ -0,0 +1,298 @@
+    command: 'jupyterlab-launcher:show',
+    category: 'Help'
+  });
+
+  app.shell.addToLeftArea(launcherWidget);
+  return launcherModel;
+}
+
+
+/**
+ * Simple encapsulation of name and callback of launcher entries.
+ */
+class LauncherItem {
+
+  readonly name: string;
+ 
+  readonly clickCallback: () => void;
+  
+  readonly imgName: string;
+ 
+  constructor(name: string, clickCallback: () => void, imgName?: string) {
+    this.name = name;
+    this.clickCallback = clickCallback;
+    if (imgName) {
+      this.imgName = imgName;
+    } else {
+      this.imgName = 'jp-Image' + name.replace(/\ /g, '');
+    }
+  }
+}
+
+
+/**
+ * LauncherModel keeps track of the path to working directory and has a list of
+ * LauncherItems.
+ */
+class LauncherModel extends VDomModel implements ILauncher {
+  items: LauncherItem[] = [];
+  path: string = 'home';
+  app: JupyterLab;
+  /**
+   * Convenience method to add a launcher with a given name and callback.
+   *
+   * This keeps plugins who wish to register themselves in the Launcher from
+   * having to import both LauncherItems and LauncherItem.
+   */
+  add(name: string, action: string, args?: JSONObject, imgName?: string) : IDisposable {
+    let clickCallback = () => { this.app.commands.execute( action, args) };
+    return this.addItem(new LauncherItem(name, clickCallback, imgName));
+  }
+
+  /**
+   * Add a new launcher and trigger re-render event for parent widget.
+   *
+   * @returns A Disposable which can be called to remove this new item from
+   * the Launcher and trigger another re-render event.
+   */
+  addItem(item: LauncherItem) : IDisposable {
+    this.items.push(item);
+    this.stateChanged.emit(void 0);
+
+    return new DisposableDelegate(() => {
+      // Remove the item form the list of items.
+      var index = this.items.indexOf(item, 0);
+      if (index > -1) {
+          this.items.splice(index, 1);
+          this.stateChanged.emit(void 0);
+      }
+    });
+  }
+
+  /**
+   * Set the path to the current working directory.
+   */
+  setDir(path: string) : void {
+    this.path = path;
+    this.stateChanged.emit(void 0);
+  }
+
+  /**
+   * Set the JupyterLab application this launcher will use when executing
+   * commands.
+   */
+  setApp(app: JupyterLab) : void {
+    this.app = app;
+  }
+}
+
+
+/**
+ * A virtual-DOM-based widget for the Launcher.
+ */
+class LauncherWidget extends VDomWidget<LauncherModel> {
+
+  /**
+   * Construct a new launcher widget.
+   */
+  constructor() {
+    super();
+    this.addClass(LAUNCHER_CLASS);
+  }
+
+  /**
+   * Render the launcher to virtual DOM nodes.
+   */
+  protected render(): VNode | VNode[] {
+    let children : VNode[] = [];
+
+    for (let item of this.model.items) {
+      let img = h.span({className: item.imgName + ' ' + IMAGE_CLASS});
+      let text = h.span({className:  TEXT_CLASS }, item.name);
+
+      let column = h.div({
+        className: COLUMN_CLASS,
+        'onclick': item.clickCallback
```

Comment:
OK I see how the `clickCallback` is needed here. Maybe put the command name in the `LauncherItem` API and have that be the public entry point. Then have a method or get method for getting the click callback for that command that can be used here? Mainly thinking of how we can simplify that `add` versus `addItem` method into a single one that just takes a `LauncherItem` instance.